### PR TITLE
Fixes #1354

### LINKED ISSUE #1354: Compile error when compiling with -Werror using gcc 12.0.1
## System information
Cross compiling media-driver 22.2.1 using gcc 12-20220220

### Describe the current behavior
Fails compile with:
- `warning: dangling pointer 'address2Counter' to 'ctr' may be used [-Wdangling-pointer=]`
- `warning: argument 1 range [18446744071562067968, 18446744073709551615] exceeds maximum object size 9223372036854775807 [-Walloc-size-larger-than=]`

### Describe the expected behavior
Compiles correctly with gcc 11.2.0
Using the following compile options allows these errors to pass as warnings:
- `export CXXFLAGS="${CXXFLAGS} -Wno-error=alloc-size-larger-than= -Wno-error=dangling-pointer"`

## build log

```
[43/942] Building CXX object media_driver/CMakeFiles/iHD_drv_video_OBJ.dir/agnostic/common/cm/cm_kernel_ex.cpp.o
In file included from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_common/agnostic/common/os/mos_utilities.h:138,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_common/agnostic/common/os/mos_os.h:32,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/os/mos_context.h:30,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/linux/common/ddi/media_libva_common.h:37,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/linux/common/cm/ddi/media_libva_cm.h:30,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/linux/common/cm/hal/cm_def_os.h:41,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/cm/cm_def.h:30,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/cm/cm_kernel.h:30,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/cm/cm_kernel_rt.h:30,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/cm/cm_kernel_ex.h:28,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/cm/cm_kernel_ex.cpp:27:
In static member function 'static _Ty* MosUtilities::MosNewArrayUtil(int32_t) [with _Ty = unsigned char; _Types = {}]',
    inlined from '_Ty* MOS_NewArrayUtil(int32_t) [with _Ty = unsigned char; _Types = {}]' at /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_common/agnostic/common/os/mos_utilities.h:185:46,
    inlined from 'virtual int32_t CmKernelEx::Initialize(const char*, const char*)' at /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/cm/cm_kernel_ex.cpp:192:22,
    inlined from 'virtual int32_t CmKernelEx::Initialize(const char*, const char*)' at /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/cm/cm_kernel_ex.cpp:70:9:
/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_softlet/agnostic/common/os/mos_utilities_next.h:106:20: warning: argument 1 range [18446744071562067968, 18446744073709551615] exceeds maximum object size 9223372036854775807 [-Walloc-size-larger-than=]
  106 |         _Ty* ptr = new (std::nothrow) _Ty[numElements]();
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/toolchain/x86_64-libreelec-linux-gnu/include/c++/12.0.1/bits/exception_ptr.h:40,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/toolchain/x86_64-libreelec-linux-gnu/include/c++/12.0.1/exception:168,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/toolchain/x86_64-libreelec-linux-gnu/include/c++/12.0.1/ios:39,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/toolchain/x86_64-libreelec-linux-gnu/include/c++/12.0.1/ostream:38,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/toolchain/x86_64-libreelec-linux-gnu/include/c++/12.0.1/iostream:39,
                 from /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/linux/common/cm/hal/cm_def_os.h:34:
/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/toolchain/x86_64-libreelec-linux-gnu/include/c++/12.0.1/new: In member function 'virtual int32_t CmKernelEx::Initialize(const char*, const char*)':
/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/toolchain/x86_64-libreelec-linux-gnu/include/c++/12.0.1/new:142:26: note: in a call to allocation function 'void* operator new [](std::size_t, const std::nothrow_t&)' declared here
  142 | _GLIBCXX_NODISCARD void* operator new[](std::size_t, const std::nothrow_t&) _GLIBCXX_USE_NOEXCEPT
      |                          ^~~~~~~~
[95/942] Building CXX object media_driver/CMakeFiles/iHD_drv_video_OBJ.dir/agnostic/common/codec/hal/codechal_encoder_base.cpp.o
/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/codec/hal/codechal_encoder_base.cpp: In member function 'MOS_STATUS CodechalEncoderState::ReadCounterValue(uint16_t, EncodeStatusReport*)':
/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/codec/hal/codechal_encoder_base.cpp:3941:43: warning: dangling pointer 'address2Counter' to 'ctr' may be used [-Wdangling-pointer=]
 3941 |     encodeStatusReport->HWCounterValue.IV = *(++address2Counter);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/codec/hal/codechal_encoder_base.cpp:3926:18: note: 'ctr' declared here
 3926 |         uint32_t ctr[4] = { 0 };
      |                  ^~~
/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/codec/hal/codechal_encoder_base.cpp:3937:48: warning: dangling pointer 'address2Counter' to 'ctr' may be used [-Wdangling-pointer=]
 3937 |     encodeStatusReport->HWCounterValue.Count = *address2Counter;
      |                                                ^~~~~~~~~~~~~~~~
/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.2.1/media_driver/agnostic/common/codec/hal/codechal_encoder_base.cpp:3926:18: note: 'ctr' declared here
 3926 |         uint32_t ctr[4] = { 0 };
      |                  ^~~
```

### PR DESCRIPTION
**Pull request by pingliwa**
Fixes #1354. It is to fix the compiling issue for gcc12.0.1.
[-Werror=address] [-Werror=dangling-pointer=][-Werror=alloc-size-larger-than=]

Signed-off-by: Wang, Pingli <pingli.wang@intel.com>

### CODE REVIEW

**Review by XinfengZhang — APPROVED** [review]
LGTM

**Review by FurongZhang — APPROVED** [review]

**Comment by heitbaum** [comment]
👍 Successful test builds with:

Executing (target): cmake -GNinja -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_TOOLCHAIN_FILE=/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/toolchain/etc/cmake-x86_64-libreelec-linux-gnu.conf -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=MinSizeRel -DBUILD_CMRTLIB=OFF -DENABLE_NONFREE_KERNELS=ON -DBUILD_KERNELS=ON -DBYPASS_MEDIA_ULT=yes /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.3.0
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0

and 

Executing (target): cmake -GNinja -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_TOOLCHAIN_FILE=/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/toolchain/etc/cmake-x86_64-libreelec-linux-gnu.conf -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=MinSizeRel -DBUILD_CMRTLIB=OFF -DENABLE_NONFREE_KERNELS=ON -DBUILD_KERNELS=ON -DBYPASS_MEDIA_ULT=yes /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-11.0-devel/build/media-driver-22.3.0
-- The C compiler identification is GNU 12.0.1
-- The CXX compiler identification is GNU 12.0.1

**Review by hye5 — APPROVED** [review]

File: media_driver/linux/common/codec/ddi/media_ddi_decode_vp8.cpp [hunk 1]
```diff
@@ -235,14 +235,13 @@ VAStatus DdiDecodeVP8::ParsePicParams(
     codecPicParams->ucUvModeProbs[0]            = picParam->uv_mode_probs[0];
     codecPicParams->ucUvModeProbs[1]            = picParam->uv_mode_probs[1];
     codecPicParams->ucUvModeProbs[2]            = picParam->uv_mode_probs[2];
-    if (codecPicParams->ucMvUpdateProb[0] && picParam->mv_probs[0])
```
Comment (dvrogozh) [inline media_driver/linux/common/codec/ddi/media_ddi_decode_vp8.cpp:238]: what was the particular compilation warning here and why did you drop these checks?

Comment (pingliwa) [reply media_driver/linux/common/codec/ddi/media_ddi_decode_vp8.cpp:238]: > what was the particular compilation warning here and why did you drop these checks?

It is [-Werror=address] .
codecPicParams->ucMvUpdateProb and picParam->mv_probs are array 2 dimensi as below.
    _uint8_t             ucMvUpdateProb[2][CODEC_VP8_MVP_COUNT];      // 2 for decoding row and column components
    uint8_t mv_probs[2][19];_
codecPicParams->ucMvUpdateProb[0], picParam->mv_probs[0], codecPicParams->ucMvUpdateProb[1] and picParam->mv_probs[1] could never be nullptr.

File: media_driver/linux/common/codec/ddi/media_ddi_decode_vp8.cpp [hunk 2]
```diff
@@ -235,14 +235,13 @@ VAStatus DdiDecodeVP8::ParsePicParams(
...
+
     {
```
Comment (StevenLi01) [inline media_driver/linux/common/codec/ddi/media_ddi_decode_vp8.cpp:239]: remove the brace as well?

Comment (pingliwa) [reply media_driver/linux/common/codec/ddi/media_ddi_decode_vp8.cpp]: Fixed.

File: media_driver/linux/common/codec/ddi/media_ddi_decode_vp8.cpp (hunk 1 above)
Comment (dvrogozh) [reply media_driver/linux/common/codec/ddi/media_ddi_decode_vp8.cpp:238]: ok, thx.

**Review by dvrogozh — APPROVED** [review]

**Review by StevenLi01 — APPROVED** [review]
LGTM

**Comment by dankmolot** [comment]
Successfully built on Fedora 36 with gcc 12.1.1

**Pull request merged**